plasmapy/particles/tests/test_particle_collections.py

```diff
@@ -0,0 +1,302 @@
+"""Tests for particle collections."""
+
+import astropy.units as u
+import pytest
+
+from plasmapy.particles import alpha, electron, neutron, proton
+from plasmapy.particles.atomic import atomic_number
+from plasmapy.particles.exceptions import InvalidParticleError
+from plasmapy.particles.nuclear import nuclear_reaction_energy
+from plasmapy.particles.particle_class import (
+    CustomParticle,
+    DimensionlessParticle,
+    Particle,
+)
+from plasmapy.particles.particle_collections import ParticleList
+
+custom_particle = CustomParticle(mass=1e-25 * u.kg, charge=1e-18 * u.C)
+dimensionless_particle = DimensionlessParticle(mass=1.25, charge=1.58)
+
+
+attributes = [
+    "charge",
+    "half_life",
+    "integer_charge",
+    "mass",
+    "mass_energy",
+]
+
+
+@pytest.fixture
+def various_particles():
+    """A sample `ParticleList` with several different valid particles."""
+    return ParticleList(
+        [
+            "H",
+            "He",
+            "e-",
+            "alpha",
+            "tau neutrino",
+            CustomParticle(mass=3 * u.kg, charge=5 * u.C),
+            CustomParticle(),
+            CustomParticle(mass=7 * u.kg),
+            CustomParticle(charge=11 * u.C),
+        ]
+    )
+
+
+def _everything_is_particle_or_custom_particle(iterable):
+    """
+    Test that every object in an iterable is either a `Particle` instance
+    or a `CustomParticle` instance.
+    """
+    return all([isinstance(p, (Particle, CustomParticle)) for p in iterable])
+
+
+@pytest.mark.parametrize(
+    "args",
+    [
+        (),
+        ([electron]),
+        ([electron, proton]),
+        ([electron, proton, alpha]),
+        (["e-", "e+"]),
+        ([electron, "e-"]),
+        ([custom_particle]),
+        ([custom_particle, electron, "e-"]),
+    ],
+)
+def test_particle_list_membership(args):
+    """
+    Test that the particles in the `ParticleList` match the particles
+    (or particle-like objects) that are passed to it.
+    """
+    particle_list = ParticleList(args)
+    for arg, particle in zip(args, particle_list):
+        assert particle == arg
```

Comment:
Just thinking about this: this would work like `particle == arg and particle is not arg`? As in, you don't hold references to particle objects, right? (just now going to dig into the implementation!)

Reply:
The way it works is that if `ParticleList` is provided with a `Particle` or `CustomParticle`, it will hold the reference to the original particle object.  My motivation to do it this way is to be more consistent with how a regular `list` works.